crates/block-producer/src/mempool/batch_graph.rs

```diff
@@ -0,0 +1,179 @@
+    ) {
+        // Reverse lookup parent transaction batches.
+        let parents = parents
+            .into_iter()
+            .map(|tx| self.transactions.get(&tx).expect("Parent transaction must be in a batch"))
+            .copied()
+            .collect();
+
+        // Inform parents of their new child.
+        for parent in &parents {
+            self.nodes
+                .get_mut(parent)
+                .expect("Parent batch must be present")
+                .children
+                .insert(id);
+        }
+
+        // Insert transactions for reverse lookup.
+        for tx in &transactions {
+            self.transactions.insert(*tx, id);
+        }
+
+        // Insert the new node into the graph.
+        let batch = Node {
+            status: Status::InFlight,
+            transactions,
+            parents,
+            children: Default::default(),
+        };
+        self.nodes.insert(id, batch);
+
+        // New node might be a root.
+        //
+        // This could be optimised by inlining this inside the parent loop. This would prevent the
+        // double iteration over parents, at the cost of some code duplication.
+        self.try_make_root(id);
+    }
+
+    /// Removes the batch and all of its descendents from the graph.
+    pub fn purge_subgraph(&mut self, id: BatchJobId) -> Vec<(BatchJobId, Vec<TransactionId>)> {
+        let mut removed = Vec::new();
+
+        let mut to_process = vec![id];
+
+        while let Some(batch_id) = to_process.pop() {
+            // Its possible for a node to already have been removed as part of this subgraph
+            // removal.
+            let Some(batch) = self.nodes.remove(&batch_id) else {
+                continue;
+            };
+
+            // All the child batches are also removed so no need to check
+            // for new roots. No new roots are possible as a result of this subgraph removal.
+            self.roots.remove(&batch_id);
+
+            for transaction in &batch.transactions {
+                self.transactions.remove(transaction);
+            }
+
+            // Inform parent that this child no longer exists.
+            //
+            // The same is not required for children of this batch as we will
+            // be removing those as well.
+            for parent in &batch.parents {
+                // Parent could already be removed as part of this subgraph removal.
+                if let Some(parent) = self.nodes.get_mut(parent) {
+                    parent.children.remove(&batch_id);
+                }
+            }
+
+            to_process.extend(batch.children);
+            removed.push((batch_id, batch.transactions));
+        }
+
+        removed
+    }
+
+    /// Removes a set of batches from the graph without removing any descendents.
+    ///
+    /// This is intended to cull completed batches from stale blocks.
+    pub fn remove_stale(&mut self, batches: Vec<BatchJobId>) -> Vec<TransactionId> {
```

Comment:
Similar comment as the one for transaction graph: maybe this could be `remove_committed()`.